src/content/blog/2023_qlasskit.md

```diff
@@ -0,0 +1,74 @@
+---
+title: Qlasskit - a bridge between Python and Quantum algorithms
+author: Davide Gessa
+day: 29
+month: 11
+year: 2023
+---
+
+Traditionally, creating quantum circuits requires specialized knowledge in quantum programming. This requirement holds true when encoding a classical algorithm inside a quantum circuit, for instance, for an oracle or a black-box component of a quantum algorithm. This often becomes a time wasting job, since we almost always already have a classical implementation in a traditional high level language.
+
+**Qlasskit**, an open-source Python library developed with the support of a Unitary Fund microgrant, addresses this challenge head-on by allowing direct translation of standard Python code into invertible quantum circuits without any modification to the original code. Furthermore, *qlasskit* implements some well-known quantum algorithms and offers a comprehensive interface for implementing new ones.
+
+To illustrate *qlasskit*'s capabilities, we will demonstrate its use in performing a pre-image attack on a cryptographic hash function using Grover's search algorithm, obtaining a quadratic speedup compared to classical approaches. The beauty of *qlasskit* lies in its simplicity – you can write the entire software without needing to understand any concept of quantum computing.
+
+A pre-image attack, in cryptography, targets a hash function `h(m)` with the aim to discover an original message `m` that corresponds to a specific hash value. On a traditional computer, to perform this attack without any hints, we must run `h(m)` with every possible input (`N=2**n`).
+
+Thanks to the Grover search algorithm, we are able to find a pre-image with only `pi/2 * sqrt(N)` iterations, obtaining the quadratic speedup I mentioned before. 
+
+We write a toy hash function `hash_simp` which operates on messages composed of two 4 bit values and uses bitwise xor to create an 8 bit hash value. 
+
+![](/images/2023-qlasskit/hash_simp_code.png)
+
+The first things you can notice in this code are:
+
+- the `qlassf` decorators, indicating that the function will be translated to a quantum circuit.
+- special bit-sized types Qlist, Qint4, and Qint8. These are required as qubits are a precious resource, and we want to use as few as possible.
+- and, it is normal Python code.
+
+To see the resulting quantum circuit we can export and draw in qiskit:
+
+![](/images/2023-qlasskit/circuit_draw_code.png)
+
+And this is the resulting circuit, produced by the *qlasskit* internal compiler:
+
+![](/images/2023-qlasskit/circuit_result.png)
+
+Thanks to the fact that *qlasskit* function are standard Python functions, we can call the `original_f` to perform some kind of analysis and test on the hash function. Since the input space is tiny (it is a toy hash function), we can check if the hash function is uniform (if it maps equally to the output space).
```

Comment:
```suggestion
Thanks to the fact that *qlasskit* functions are standard Python functions, we can call the `original_f` to perform some kind of analysis and test on the hash function. Since the input space is tiny (it is a toy hash function), we can check if the hash function is uniform (if it maps equally to the output space).
```